src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,380 @@
+    std::fill(owned_data_.get(), owned_data_.get() + set_size,
+              std::numeric_limits<value_type>::signaling_NaN());
+#endif  // SPECTRE_DEBUG
+    reset_pointer_vector(set_size);
+  }
+
+  VectorImpl(size_t set_size, double value) noexcept
+      : owned_data_(set_size > 0 ? static_cast<value_type*>(
+                                       malloc(set_size * sizeof(value_type)))
+                                 : nullptr,
+                    &free) {
+    std::fill(owned_data_.get(), owned_data_.get() + set_size, value);
+    reset_pointer_vector(set_size);
+  }
+  // @}
+
+  /// Create a non-owning VectorImpl that points to `start`
+  VectorImpl(T* start, size_t set_size) noexcept
+      : BaseType(start, set_size), owning_(false) {}
+
+  /// Create from an initializer list of doubles. All elements in the
+  /// `std::initializer_list` must have decimal points
+  template <class C, Requires<cpp17::is_same_v<C, T>> = nullptr>
+  VectorImpl(std::initializer_list<C> list) noexcept
+      : owned_data_(list.size() > 0 ? static_cast<value_type*>(malloc(
+                                          list.size() * sizeof(value_type)))
+                                    : nullptr,
+                    &free) {
+    // Note: can't use memcpy with an initializer list.
+    std::copy(list.begin(), list.end(), owned_data_.get());
+    reset_pointer_vector(list.size());
+  }
+
+  /// Empty VectorImpl
+  VectorImpl() = default;
+  /// \cond HIDDEN_SYMBOLS
+  ~VectorImpl() = default;
+
+  // clang-tidy: calling a base constructor other than the copy constructor.
+  // We reset the base class in reset_pointer_vector after calling its default
+  // constructor
+  VectorImpl(const VectorImpl<T, VectorType>& rhs) noexcept;
+
+  VectorImpl(VectorImpl<T, VectorType>&& rhs) noexcept;
+
+  // This is a converting constructor. clang-tidy complains that it's not
+  // explicit, but we want it to allow conversion.
+  // clang-tidy: mark as explicit (we want conversion to VectorImpl type)
+  template <
+      typename VT, bool VF,
+      Requires<cpp17::is_same_v<typename VT::ResultType, VectorType>> = nullptr>
+  VectorImpl(const blaze::DenseVector<VT, VF>& expression) noexcept;  // NOLINT
+
+  template <typename VT, bool VF>
+  VectorImpl<T, VectorType>& operator=(
+      const blaze::DenseVector<VT, VF>& expression) noexcept;
+
+  //  declaration for copy constructor
+  VectorImpl<T, VectorType>& operator=(
+      const VectorImpl<T, VectorType>& rhs) noexcept;
+
+  // declaration for move constructor
+  VectorImpl<T, VectorType>& operator=(
+      VectorImpl<T, VectorType>&& rhs) noexcept;
+  /// \endcond
+
+  // create from a single stored value, which becomes the sole entry.
+  VectorImpl<T, VectorType>& operator=(const T& rhs) noexcept;
+
+  // @{
+  /// Set the VectorImpl to be a reference to another VectorImpl object
+  template <typename RhsVectorType>
+  void set_data_ref(gsl::not_null<RhsVectorType*> rhs) noexcept {
```

Comment:
This change is slightly more restrictive than the `static_assert` below, and allows less flexibility with the the `ResultType` template parameter. Effectively, this restricts this function to behaving as though the CRTP is always used. I have no objection to that, just wanted to make sure that the effects were clear.

Reply:
Given this is fairly low-level memory management I think the only way to correctly use this is with the tighter restriction. Or at least I think it'll go wrong more often than it'll go right otherwise :)